src/gui/qgsexternalstoragefilewidget.h

```diff
@@ -0,0 +1,198 @@
+    Q_PROPERTY( QString auth READ storageAuthConfigId WRITE setStorageAuthConfigId )
+    Q_PROPERTY( QString storageUrlExpression READ storageUrlExpressionString WRITE setStorageUrlExpression )
+
+  public:
+
+    /**
+     * \brief QgsExternalStorageFileWidget creates a widget for selecting a file or a folder.
+     */
+    explicit QgsExternalStorageFileWidget( QWidget *parent SIP_TRANSFERTHIS = nullptr );
+
+    /**
+     * Set \a storageType storage type unique identifier as defined in QgsExternalStorageRegistry or
+     * null QString if there is no storage defined.
+     * If no external storage has been defined, QgsExternalStorageFileWidget will only update file path according to
+     * selected files.
+     * \see storageType
+     * \since QGIS 3.22
+     */
+    void setStorageType( const QString &storageType );
+
+    /**
+     * Returns storage type unique identifier as defined in QgsExternalStorageRegistry.
+     * Returns null QString if there is no storage defined, only file selection.
+     * \see setStorageType
+     * \since QGIS 3.22
+     */
+    QString storageType() const;
+
+    /**
+     * Returns external storage used to store selected file names, nullptr if none have been defined.
+     * If no external storage has been defined, QgsExternalStorageFileWidget will only update file path according to
+     * selected files.
+     * \see setStorageType
+     * \since QGIS 3.22
+     */
+    QgsExternalStorage *externalStorage() const;
+
+    /**
+     * Sets the authentication configuration ID to be used for the current external storage (if
+     * defined)
+     * \since QGIS 3.22
+     */
+    void setStorageAuthConfigId( const QString &authCfg );
+
+    /**
+     * Returns the authentication configuration ID used for the current external storage (if defined)
+     * \since QGIS 3.22
+     */
+    const QString &storageAuthConfigId() const;
+
+    /**
+     * Set \a urlExpression expression, which once evaluated, provide the URL used to store selected
+     * documents. This is used only if an external storage has been defined
+     * \see setStorageType(), externalStorage()
+     * \since 3.22
+     */
+    void setStorageUrlExpression( const QString &urlExpression );
+
+    /**
+     * Returns the original, unmodified expression string, which once evaluated, provide the
+     * URL used to store selected documents. This is used only if an external storage has been defined.
+     * Returns null if no expression has been set.
+     * \see setStorageUrlExpression()
+     * \since 3.22
+     */
+    QString storageUrlExpressionString() const;
+
+    /**
+     * Returns expression, which once evaluated, provide the URL used to store selected
+     * documents. This is used only if an external storage has been defined.
+     * Returns null if no expression has been set.
+     * \see setStorageUrlExpression()
+     * \since 3.22
+     */
+    QgsExpression *storageUrlExpression() const;
```

Comment:
following the discussion here https://github.com/qgis/QGIS/pull/44533#discussion_r682179263
indeed in other places, expression are returned by pointer.
any suggestion/objection @nyalldawson